suave/cstore/engine2.go

```diff
@@ -0,0 +1,283 @@
+	return nil
+}
+
+func (a *Engine2) finalize(store *TransactionalStore2) error {
+	for _, record := range store.pendingRecords {
+		err := a.storage.InitRecord(record)
+		if err != nil {
+			// TODO: deinitialize!
+			return fmt.Errorf("confidential engine: store backend failed to initialize record: %w", err)
+		}
+	}
+
+	for _, sw := range store.pendingWrites {
+		if _, err := a.storage.Store(sw.DataRecord, sw.Caller, sw.Key, sw.Value); err != nil {
+			// TODO: deinitialize and deStore!
+			return fmt.Errorf("failed to store data: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (a *Engine2) getSession(id string) *TransactionalStore2 {
+	a.sessionsLock.Lock()
+	defer a.sessionsLock.Unlock()
+	return a.sessions[id]
+}
+
+func (a *Engine2) handleNewSession(w http.ResponseWriter, r *http.Request) {
+	id := uuid.New().String()[:7]
+	store := NewTransactionalStore2(a.storage)
+
+	a.sessionsLock.Lock()
+	a.sessions[id] = store
+	a.sessionsLock.Unlock()
+
+	fmt.Fprint(w, id)
+}
+
+func (a *Engine2) handleNewRecord(w http.ResponseWriter, r *http.Request) {
+	id := mux.Vars(r)["id"]
+
+	session := a.getSession(id)
+	if session == nil {
+		http.Error(w, "session not found", http.StatusNotFound)
+		return
+	}
+
+	// decode namespace from query. If not found, return 404
+	namespace := r.URL.Query().Get("namespace")
+	if namespace == "" {
+		http.Error(w, "namespace not found", http.StatusNotFound)
+		return
+	}
+
+	record := types.DataRecord{
+		Salt:                suave.RandomDataRecordId(),
+		DecryptionCondition: 0,
+		Version:             namespace,
+	}
+
+	// decode decrypt condition from query. If not found, default to 0
+	decryptConditionStr := r.URL.Query().Get("decryptCondition")
+	if decryptConditionStr != "" {
+		decryptCondition, err := strconv.ParseUint(decryptConditionStr, 10, 64)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to parse decrypt condition: %v", err), http.StatusBadRequest)
+			return
+		}
+		record.DecryptionCondition = decryptCondition
+	}
+
+	record, err := session.InitRecord(record)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to init record: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	fmt.Fprintf(w, "%s", hex.EncodeToString(record.Id[:]))
+}
+
+func (a *Engine2) handleRecordStorePost(w http.ResponseWriter, r *http.Request) {
+	id := mux.Vars(r)["id"]
+	key := mux.Vars(r)["key"]
+	dataId := decodeDataId(r)
+
+	session := a.getSession(id)
+	if session == nil {
+		http.Error(w, "session not found", http.StatusNotFound)
+		return
+	}
+
+	value := []byte(r.URL.Query().Get("value"))
+	session.Store(dataId, common.Address{}, key, value)
+}
+
+func (a *Engine2) handleRecordStoreGet(w http.ResponseWriter, r *http.Request) {
+	id := mux.Vars(r)["id"]
+	key := mux.Vars(r)["key"]
+	dataId := decodeDataId(r)
+
+	session := a.getSession(id)
+	if session == nil {
+		http.Error(w, "session not found", http.StatusNotFound)
+		return
+	}
+
+	val, err := session.Retrieve(dataId, common.Address{}, key)
```

Comment:
Is the empty address intentional?

Reply:
This was an argument required for the storage implementation which in turn was not used. We can do a followup PR to change the storage interface. For this PR I wanted to have the minimum possible changes to the legacy store.